hail/src/main/scala/is/hail/expr/ir/lowering/LowerBlockMatrixIR.scala

```diff
@@ -0,0 +1,155 @@
+package is.hail.expr.ir.lowering
+
+import is.hail.expr.Nat
+import is.hail.expr.ir._
+import is.hail.expr.ir.functions.GetElement
+import is.hail.expr.types.BlockMatrixSparsity
+import is.hail.expr.types.virtual._
+import is.hail.utils.{FastIndexedSeq, FastSeq}
+
+object BlockMatrixStage {
+  def empty(eltType: Type): BlockMatrixStage =
+    EmptyBlockMatrixStage(eltType)
+}
+
+case class EmptyBlockMatrixStage(eltType: Type) extends BlockMatrixStage(Array(), TInt32()) {
+  def blockContext(idx: (Int, Int)): IR =
+    throw new LowererUnsupportedOperation("empty stage has no block contexts!")
+  def blockBody(ctxRef: Ref): IR = NA(TNDArray(eltType, Nat(2)))
+  override def collectBlocks(f: IR => IR, blocksToCollect: Array[(Int, Int)]): IR = {
+    assert(blocksToCollect.isEmpty)
+    MakeArray(FastSeq(), TArray(f(blockBody(Ref("x", ctxType))).typ))
+  }
+}
+
+abstract class BlockMatrixStage(val globalVals: Array[(String, IR)], val ctxType: Type) {
+  def blockContext(idx: (Int, Int)): IR
+  def blockBody(ctxRef: Ref): IR
+
+  def collectBlocks(f: IR => IR, blocksToCollect: Array[(Int, Int)]): IR = {
+    val ctxRef = Ref(genUID(), ctxType)
+    val body = f(blockBody(ctxRef))
+    val ctxs = MakeArray(blocksToCollect.map(idx => blockContext(idx)), TArray(ctxRef.typ))
+    val bcFields = globalVals.filter { case (f, _) => Mentions(body, f) }
```

Comment:
This is going to compute the free variables of `body` for each global. Maybe `globalVals` is always small, but it's only an extra line or two to precompute `FreeVariables(body)`.

Reply:
Oops, good call!